beacon-chain/sync/data_columns_reconstruct.go

```diff
@@ -0,0 +1,187 @@
+package sync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	cKzg4844 "github.com/ethereum/c-kzg-4844/bindings/go"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/core/feed"
+	statefeed "github.com/prysmaticlabs/prysm/v5/beacon-chain/core/feed/state"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/core/peerdas"
+	fieldparams "github.com/prysmaticlabs/prysm/v5/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/interfaces"
+	ethpb "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/sirupsen/logrus"
+)
+
+// recoverBlobs recovers the blobs from the data column sidecars.
+func recoverBlobs(
+	dataColumnSideCars []*ethpb.DataColumnSidecar,
+	columnsCount int,
+	blockRoot [fieldparams.RootLength]byte,
+) ([]cKzg4844.Blob, error) {
+	recoveredBlobs := make([]cKzg4844.Blob, 0, fieldparams.MaxBlobsPerBlock)
+
+	for blobIndex := 0; blobIndex < fieldparams.MaxBlobsPerBlock; blobIndex++ {
+		start := time.Now()
+
+		cellsId := make([]uint64, 0, columnsCount)
+		cKzgCells := make([]cKzg4844.Cell, 0, columnsCount)
+
+		for _, sidecar := range dataColumnSideCars {
```

Comment:
are the sidecars assumed to be inorder ? otherwise you would have something looking very different

Reply:
There is not such an assumption.

However, the order in `cellIsd` must be the same than the order in `cKzgCells`, since they are given to
```go
recoveredCells, err := cKzg4844.RecoverAllCells(cellsId, cKzgCells)
```

The order itself does not matter, but the order between `cellsId` and `cKzgCells` does.
(And this order is respected in the current implementation.)